src/cascade/model/model_writer.py

```diff
@@ -0,0 +1,356 @@
+"""
+Writes a Model to a Dismod File.
+"""
+from math import nan, inf
+from numbers import Real
+
+import numpy as np
+import pandas as pd
+
+from cascade.core import getLoggers
+from cascade.dismod.constants import DensityEnum, WeightEnum, MulCovEnum
+
+CODELOG, MATHLOG = getLoggers(__name__)
+
+
+class ModelWriter:
+    """This layer over the Dismod File is responsible for translation between
+    absolute numbers in the Model and SQL-indexed numbers.
+
+    This has an API that the client has to follow. Functions need to be called
+    in order.
+
+    For naming in the DismodFile, this chooses names based on the random
+    fields.
+
+    Assumptions this ModelWriter makes, on top of what Dismod-AT requires:
+
+     * Locations have a ``c_location_id`` member that ``maps node_id``
+       to ``location_id``.
+     * Rates and integrands are always in the same order.
+    """
+
+    def __init__(self, session, extremal_age_time):
+        """
+
+        Args:
+            session (Session): The Dismod-AT Session into which this writes.
+            extremal_age_time (Tuple(Set(int),Set(int))): a set of ages and
```

Comment:
Can we really not defer this so the user doesn't have to think about it? This must be available when write_ages_and_times is called, right? It could be collected at that point?

Reply:
Or the intention here is that there will be the need to add int extremes beyond what's in the model and data?

Reply:
The user doesn't call this. The session does. And, yes, it sucks. Thank Brad. I didn't catch that this needed to be done until implementing the fifth user example from Brad, and something broke because it needed a fake age_id put in to account for the data age range. The point is to get the extra age from the **data**, not the model. So it can't happen when writing ages and times.

Reply:
Hmm. Ok. Wait does the model writer not also write the data?

Reply:
Because data isn't part of the model? And this is just writing the model... Ok

Reply:
The data isn't part of the model. That's totally it. It's a stats definition of what a model is.